### ISSUE TITLE
Quotient graphs - option to label block nodes via a mapping on creation

### ISSUE THREAD

**Issue by sr-murthy** [networkx/networkx#3837]
There doesn't seem to be an option to set the labels of the block nodes of a quotient graph via a mapping passed in to the creation method [`networkx.quotient_graph`](https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.minors.quotient_graph.html#networkx.algorithms.minors.quotient_graph). There is a `relabel` boolean that labels the block nodes as non-negative integers, but there is no way of controlling this if the block nodes have to be labelled according to some mapping.

This is an example to illustrate the problem. I have a (directed) strongly connected graph `G` with 21 nodes

    [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40]

There is an equivalence relation `nrel` on these nodes which produces the following six blocks

    [0, 2, 4, 6, 8]      #B1
    [16, 10, 12, 14]     #B2
    [24, 18, 20, 22]     #B3
    [32, 26, 28, 30]     #B4
    [34, 36]             #B5
    [40, 38]             #B6

and an edge relation `brel` on the blocks given by

    >>> brel = lambda B, C: any(e in G.edges for e in product(B, C))

(A simple extension of the edge relation of `G` to the blocks of `Q` - there is an edge from block `B` to `C` if there is an edge in `G` from a node in `B` to a node in `C`).

I create the quotient `Q` using

    >>> Q = nx.quotient_graph(G, nrel, edge_relation=brel)

The nodes of `Q` are now these `frozenset` objects

    >>>Q.nodes
    >>> NodeView((frozenset({34, 36}), frozenset({16, 10, 12, 14}), frozenset({32, 26, 28, 30}), frozenset({40, 38}), frozenset({0, 2, 4, 6, 8}), frozenset({24, 18, 20, 22})))

But I would like to pass a mapping to `nx.quotient_graph` which sets the labels of the block nodes according to this mapping

    >>> block_relabelling = {frozenset({0, 2, 4, 6, 8}): 2, frozenset({10, 12, 14, 16}): 3, frozenset({18, 20, 22, 24}): 4, frozenset({[26, 28, 30, 32]}): 5, frozenset({34, 36}): 6, frozenset({38, 40}): 7}
    >>> nx.relabel_nodes(Q, block_relabelling, copy=False)
    >>> sorted(Q.nodes)
    >>> [2, 3, 4, 5, 6, 7]

It would be convenient if this could be done via an optional argument during the creation of the quotient graph, i.e. if [`networkx.quotient_graph`](https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.minors.quotient_graph.html#networkx.algorithms.minors.quotient_graph) had an optional argument `block_relabel` (or something like that) which accepted a mapping from blocks to labels and applied the mapping to the blocks on creation of the quotient graph.

**Comment by sr-murthy** [comment]
@dschult If you think this would be useful, I would be happy to submit a PR.

**Comment by dschult** [comment]
Is the advantage of this that you can choose the node representation of each block?
The relabel argument conveniently renames the nodes as integers.
What does this new feature offer that the relabel argument doesn't?

And just to be sure I understand, this would allow the following in one call:
```python
Q = nx.quotient_graph(G, nrel, edge_relation=brel)
nx.relabel_nodes(Q, block_relabelling, copy=False)
```

**Comment by sr-murthy** [comment]
Yes, it allows the labelling of the block nodes in the quotient to be customised, because in certain contexts the sequence of consecutive non-negative integer labels produced by the `relabel` doesn't really make sense in terms of the function which is the basis for the equivalence relation used to partition the nodes in the original graph `G` - say, `f` is a function on the nodes of `G`, then of course there is a natural equivalence relation on the nodes of `G` given by: `x rel y` if and only if `f(x)= f(y)`. Then `rel` partitions `G` into equivalence classes `C_1,...,C_k`, and it is natural to label these classes by the images under `f`: `f(C_1),...,f(C_k)`. I want to use `f(C_1),...,f(C_k)` as the labels for the blocks of `Q`, not `0,...,k - 1`.

**Comment by dschult** [comment]
How about we abuse the ```relabel``` argument?
If it is True, relabel as it currently does. If False, use frozensets. If not True or False, it should be a dict with frozensets as keys pointing to values that are a node object for that set. 

Are there potential difficulties or confusions with this approach?

**Comment by sr-murthy** [comment]
I think that sounds OK to me. So the behaviour of `quotient_graph` when `relabel` is `True` or `False` would be unchanged (map blocks to non-negative integers if `True`, otherwise use frozen sets of the block elements), but a dict would cause its mapping to be applied instead.

By the way, when `relabel` is `True` what mapping is applied to produce the non-negative integers as labels for the blocks? Looking at the relabelled nodes of my quotient graph `Q`, in my example graph `G` (with nodes  `[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40]`) here is the mapping I see

    {34, 36} -> 0
    {10, 12, 14, 16} -> 1
    {26, 28, 30, 32} -> 2
    {38, 40} -> 3
    {0, 2, 4, 6, 8} -> 4
    {18, 20, 22, 24} -> 5

This is a specific mapping, but why this one, and not one where, say, `{34, 36} -> 3`, for example? My suggestion for `relabel` in calling `quotient_graph` is simply to allow the user to determine this mapping.

I don't think there would be any issues with this - after all, the blocks can be relabelled according to any mapping using `nx.relabel`.

**Comment by sr-murthy** [comment]
So if it's OK with you, I will submit a PR.

**Comment by sr-murthy** [comment]
BTW passing a mapping to `relabel` in the suggestion above would also work identically to passing the original function `f` associated with the equivalence relation `rel`on the nodes of `G`. So `rel` is something like

    rel = lambda x, y: f(x) == f(y)

where `f` is a function on the nodes of `G`. As this function `f` has to exist for `rel` you could also pass `f` as the value of `relabel` (when calling `nx.quotient_graph`), e.g.

    nx.quotient_graph(G, rel, edge_relation=block_rel, relabel=f)

and what this would do is to apply `f` on a singly chosen element of each block. In my example above the mapping defined by `f` on `G` is given by

    f({0, 2, 4, 6, 8}) = 2
    f({10, 12, 14, 16}) = 3
    f({18, 20, 22, 24}) = 4
    f({26, 28, 30, 32}) = 5
    f({34, 36}) = 6
    f({38, 40}) = 7

Then `nx.quotient_graph(G, rel, edge_relation=block_rel, relabel=f)` could be implemented by creating the quotient with its blocks using `rel` and then labelling the blocks using `f` as:

    block {0, 2, 4, 6, 8}  -> f(0) = 2
    block {10, 12, 14, 16} -> f(10) = 3
    block {18, 20, 22, 24} -> f(18) = 4
    block {26, 28, 30, 32} -> f(26) = 5
    block {34, 36}         -> f(34) = 6
    block {38, 40}         -> f(38) = 7 

Here `0, 10, 18, 26, 34, 38` were chosen from the blocks by taking the first element of each block in sorted order.

**Comment by sr-murthy** [comment]
A clarification on your earlier question/suggestion:

>How about we abuse the relabel argument?

>If it is True, relabel as it currently does. If False, use frozensets. If not True or False, it should be a dict with frozensets as keys pointing to values that are a node object for that set.

This would mean three different values for `relabel` with three different actions. I think this is more complex than having a default of `None`, which would keep the blocks as frozensets (as is the current behaviour when `relabel=False`), or a `dict` mapping of blocks to labels. This way, there would be just two actions for `relabel`.

**Comment by sr-murthy** [comment]
https://github.com/networkx/networkx/pull/3839

**Comment by dschult** [comment]
I like what you've got here (haven't looked at the code yet).

I'm a little concerned about backward compatibility. If we change too much, people who have ```relabel=True``` in their code will have their code break when they upgrade. 

The argument being a function seems a little clumsy -- we are asking the user to provide the same function twice. Does it need to be the same function? If not, fair enough. If so, perhaps we can use that fact along with the ```relabel=True``` option to signal to use that function.

**Comment by sr-murthy** [comment]
On backwards compatibility I agree that it is important to have, and I am more happy to follow your original suggestion and keep `relabel` but add the option for it to accept dicts and functions for relabelling, which is my contribution. The concern which this approach is that you are adding to the complexity to the step of processing the different possibilities for `relabel`

https://github.com/networkx/networkx/blob/master/networkx/algorithms/minors.py#L278

On your second point, the function you're referring to is not the same as what the `partition` argument is supposed to accept. The docstring (https://github.com/networkx/networkx/blob/master/networkx/algorithms/minors.py#L64) says

    partition : function or list of sets
        If a function, this function must represent an equivalence
        relation on the nodes of `G`....

So this would be something like

    rel = lambda x, y: f(x) == f(y)

So `partition` here would be `rel` not the inner function `f`, which is the function on the nodes of `G` which is the basis for the equivalence relation. It is this inner function which I've used in the example above to do the relabelling (passed in for `labels=`), but you cannot programmatically recover this `f` function from `partition=rel`, and `rel` isn't useful for relabelling.

So `rel` and `f` are different, but related of course.

**Comment by MridulS** [comment]
I believe this is fixed in https://github.com/networkx/networkx/pull/4183. I'll go ahead and close this :)
